### ISSUE TITLE
Manuscript '21 | How to Train BERT with an Academic Budget

### ISSUE THREAD

**Issue by vycezhong** [vycezhong/read-papers#173]
https://peteriz.github.io/posts/papers/academic-budget-bert.pdf

推特看到的. 感觉有意思. 用8个Titan-V (12GB)的GPU 24h训练BERT

**Comment by vycezhong** [comment]
training tricks:

1. 只训练max_seq_len = 128. 

2. single-sequence training. 不用NSP. 

3. 减少validation次数.

4. 采用BERT-large模型. 

5. optimizer: AdamW. \beta_1 = 0.9, \beta_2 = 0.98, \eps=1e-6, wd=0.01, dropout=0.1

6. 使用[DeepSpeed](https://github.com/microsoft/DeepSpeed). 他们有不少优化，比如fuse_adam, fuse_transformer等. 

7. mixed precision training. 

8. in-memory dataset. 优化I/O

一些调参技巧:

1. bs: 用gradient-accumulation来增大batch size. 很明智的做法，这能减少通信次数. 使用了4k, 8k, 16k. （不过这么大bs确定AdamW能行？

2. lr: linear learning rate scheduler. lr = {5e-4, 1e-3, 2e-3}

3. warmup proportion(wu):  wu={0%, 2%, 4%, 6%}. 和LAMB #3 差不多. bs=4k差不多是1/40. 8k是1/20.


最后结果
![image](https://user-images.githubusercontent.com/25879526/115100748-3d177380-9f71-11eb-842c-ef96a55c0071.png)
nb. 

![image](https://user-images.githubusercontent.com/25879526/115100751-430d5480-9f71-11eb-95ed-0215717ecb0a.png)

这个表中原BERT loss这么高有点奇怪. 一般phase1跑完也有1.7. 


![image](https://user-images.githubusercontent.com/25879526/115100781-954e7580-9f71-11eb-9965-f57e532e4832.png)
nb. 确实训练了bert. 

没跑SQuAD. 不知道为什么.
